### PR TITLE
Adds Capitalism Firing Pins

### PR DESCRIPTION
**Pull request by MMMiracles**
## About The Pull Request

Adds a paywall firing pin, a firing pin that a user must first swipe an ID on with a valid bank account to link with. Using the pin in your hand lets you set the pin to either per-shot payment or one-time payment. When a user with a valid bank account attempts to fire the gun, they'll get a prompt asking them if they want to pay the fee. Per-shot users will only get this prompt once as a general warning that each shot is going to cost them money. You can examine the gun to figure out whether its a per-shot or one-time payment and how much said payment is.

The crates for the pins cost 500 more than generic pins due to the fact these can act as shittier DNA pins if the price is set high enough. I'd rather not raise the price too much though so people actually use them for gun cargo/security handouts.

image

## Why It's Good For The Game

Someone bought me ARMA 3 and getting a big on-screen prompt mid-combat about how I should buy the MARKSMAN DLC to use that weapon's specific scope was really funny to me and I thought it'd be funny to have in this game.

## Changelog
:cl: MMMiracles
add: Paywall firing pins! Force people to properly buy your guns before they try to use them on you! Found where regular firing pins are sold.
/:cl:

### CODE REVIEW

**Comment by Shapsy** [comment]
I pulled my own gun, put a quarter in it, and fired back. The bullet lodged in a U.S.P.S. mailbox less than a foot from his head. I shot the mailbox again, on purpose.

**Comment by cy294** [comment]
You need to buy the Robust™ DLC Package before you can use this laser gun.

**Comment by dr-bee** [comment]
Guns that charge your account for every shot fired when?

**Comment by optimumtact** [comment]
bad idea

**Comment by optimumtact** [comment]
changed my mind

**Comment by Difrit** [comment]
I like the overall idea of this, but I'd think it could have more of an impact with some different changes.

An alternate type of capitalism pin should have the amount of money they charge set by interacting with the pin before putting it in a weapon. When the weapon is fired, it tries to draw said amount of money from the worn ID. If it can, all is well and good, it takes the money and drops it into a department account, god knows which. If it doesn't, the gun doesn't fire and you get an "I'm poor..." negative moodlet.

**Comment by Tlaltecuhtli** [comment]
shouldnt the pin_owner be the bank account instead of the id itself?

**Comment by Shadowflame909** [comment]
What if you had to pay per shot

And you could set the payment amount?

I see some good adminbus coming out of a pulse rifle that you have to pay 100k to shoot each time

**Comment by fludd12** [comment]
In line with @dr-bee and @Shadowflame909's comments, I'd like to see a "pay-per-shot" pin that, if you don't have enough to pay, makes the gun backfire a-la Clumsy. Proper ANCAP firing pins.

**Comment by ExcessiveUseOfCobblestone** [comment]
you can pay 100K to use this weapon OR steal from sec... the choice is yours, traveler...

**Comment by Rowelll** [comment]
unrelated, but buy marksman dlc - it's worth it

**Comment by OverlyGeneric** [comment]
gun cargo gun cargo

**Comment by Shadowflame909** [comment]
At the very least. I hope there's a variant of these that are non-removable.

It'd be pretty lame to see someone just buy/use another firepin for your guns.

Also. Does the money just disappear or does it get paid into someone's account?

If its the second one. I recommend paying 80% of that balance into someone's account. Claiming the 20% was used for blue space transaction reasons. Gotta snipe those exploits before they exist.

File: code/modules/projectiles/pins.dm [hunk 1]
```diff
@@ -241,9 +242,11 @@
 			return TRUE
 		if(H.get_bank_account())
 			credit_card_details = H.get_bank_account()
-		if(credit_card_details)
+		if(credit_card_details && !active_prompt)
 			var/license_request = input("Do you wish to pay [payment_amount] for license of this [gun.name]?", "License Purchase") as null|anything in list("Yes","No")
```
Comment (Akrilla) [inline code/modules/projectiles/pins.dm]: You should probably remove the null from `as null|anything` to remove the cancel button, as it's basically just acting as a another No button anyway. Saves the `if(!license_request)` below too.

**Comment by Tlaltecuhtli** [comment]
can you make them so you can set them to pay every time it shoots?

**Comment by MMMiracles** [comment]
do you have a reasonable way to do that which doesn't involve majority of people setting the per-fire cost extremely high (without putting a hard cap on it) to bankrupt anyone who uses it after 1-2 shots while giving fair warning to the person using it?

File: code/modules/projectiles/pins.dm (hunk 1 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/projectiles/pins.dm]: just use alert instead wtf

**Comment by Shadowflame909** [comment]
Why don't you want to bankrupt people

Corporate Greed knows no bounds

But in Earnest. Can't you just put a warning on it before you first fire. Similar to the BoH warning. This'll bankrupt you. Do you accept potentially losing all your money? If you hit yes. You're added to the list of people able to use the gun. If you press no. You can't.

File: code/modules/projectiles/pins.dm (hunk 1 above)
Comment (MMMiracles) [reply code/modules/projectiles/pins.dm]: i didn't realize alert existed as an option before-hand, what does it actually do in comparison to input, or is it just a standardization thing?

**Comment by Tlaltecuhtli** [comment]
@MMMiracles  show price tag on examinate?

**Comment by MMMiracles** [comment]
i added per-shot payment with a proper prompt to the user about said cost, examining the gun now also tells you if the gun is per-shot or one-time payment with the cost shown as well.

File: code/modules/projectiles/pins.dm [hunk 2]
```diff
@@ -249,28 +257,30 @@
...
+			if(multi_payment && credit_card_details)
+				if(credit_card_details.adjust_money(-payment_amount))
+					pin_owner.registered_account.adjust_money(payment_amount)
+					return TRUE
+				to_chat(user, "<span class='warning'>ERROR: User balance insufficent for successful transaction!</span>")	
+				return FALSE	
+			return TRUE				
 		if(credit_card_details && !active_prompt)
-			var/license_request = alert(usr, "Do you wish to pay [payment_amount] for license of this [gun.name]?", "License Purchase", "Yes", "No")
+			var/license_request = alert(usr, "Do you wish to pay [payment_amount] credit[( payment_amount > 1 ) ? "s" : ""] for [( multi_payment ) ? "each shot of [gun.name]" : "usage license of [gun.name]"]?", "Weapon Purchase", "Yes", "No")
```
Comment (PKPenguin321) [inline code/modules/projectiles/pins.dm]: im skimming here but does this go off every time you try to shoot? that's funny if so but also what happens if you keep this prompt open, throw the guy away somewhere, then accept the prompt while not being near the gun?

**Comment by MMMiracles** [comment]
There's a check higher up that looks to see if you've already bought the permit so you will only see the prompt once 

I should probably add a range check for that oops

File: code/modules/projectiles/pins.dm (hunk 2 above)
Comment (MMMiracles) [reply code/modules/projectiles/pins.dm]: There's a check higher up that looks to see if you've already bought the permit so you will only see the prompt once

I should probably add a range check for that oops

**Comment by Shadowflame909** [comment]
That's an epic gamer move MMMiracles

Cheers and speed merge.

Now to pray as the chaplain for a pulse rifle that costs 100k to fire per shot

File: code/modules/projectiles/pins.dm (hunk 1 above)
Comment (Akrilla) [reply code/modules/projectiles/pins.dm]: input supports actual text if you don't specific that list I think? Either works, but best to remove that cancel button either way.

**Comment by Shapsy** [comment]
Honestly the gun beeping "YOUR ACCOUNT HAS BEEN CHARGED 5 CREDITS" every time you shoot it successfully sounds funny as well

**Comment by actioninja** [comment]
stop bikeshedding

**Comment by adamkad1** [comment]
does it cost the owner of the account to fire the gun?

**Pull request merged**